src/Adapter/Address/CommandHandler/EditManufacturerAddressHandler.php

```diff
@@ -0,0 +1,110 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Address\CommandHandler;
+
+use Address;
+use PrestaShop\PrestaShop\Adapter\Address\AbstractManufacturerAddressHandler;
+use PrestaShop\PrestaShop\Core\Domain\Address\Command\EditManufacturerAddressCommand;
+use PrestaShop\PrestaShop\Core\Domain\Address\CommandHandler\EditManufacturerAddressHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Address\Exception\AddressException;
+use PrestaShopException;
+
+/**
+ * Handles command which edits manufacturer address
+ */
+final class EditManufacturerAddressHandler extends AbstractManufacturerAddressHandler implements EditManufacturerAddressHandlerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(EditManufacturerAddressCommand $command)
+    {
+        $addressId = $command->getAddressId();
+        $address = $this->getAddress($addressId);
+        $this->populateAddressWithData($address, $command);
+
+        try {
+            if (!$address->update()) {
+                throw new AddressException(
+                    sprintf('Cannot update address with id "%s"', $address->id)
+                );
+            }
+        } catch (PrestaShopException $e) {
+            throw new AddressException(
+                sprintf('Cannot update address with id "%s"', $address->id)
+            );
+        }
+
+        return $addressId;
+    }
+
+    /**
+     * Populates Address object with given data
+     *
+     * @param $address
+     * @param EditManufacturerAddressCommand $command
+     */
+    private function populateAddressWithData(Address $address, $command)
+    {
+        if (null !== $command->getManufacturerId()) {
+            $address->id_manufacturer = $command->getManufacturerId();
```

Comment:
@matks,
`EditManufacturerAddressCommand` accepts parameter `manufacturerId` which can be `int` in case you want to update it or `null` if you want to reset it.

The problem is when we have `Address` with lets say Manufacturer id `2` and we only want to update Addresse's city (or any other field except manufacturer). So, we create command like so:

```php
$command = new EditManufacturerAddressCommand(2);
$command->setCity('New city');

// when dispatching $command we don't expect manufacturer id to be changed
// as we did not set it explicitly.
$bus->handle($command);
```

However, in this command Manufacturer id is `null` as we didn't set it. But when handling command, we never know if `null` was set intentionally (to reset addresse's manufacturer) or left untouched as both values ends up being `null`. Thus, we don't know whether Addresse's manufacturer Id should be reset or kept as is.

Do you have any ideas of how we could tell handler if manufacturer's id for address has changed or not?

Reply:
First question: so this means we can reset an address Manufacturer Id ? I'd say yes, since you want to handle this usecase.

I think we can acknowledge the intentional passing of `null` by keeping a flag:
```
private $manufacturerIdShouldBeReset = false;

public function setManufacturerId($manufacturerId)
    {
        if (null !== $manufacturerId) {
            $manufacturerId = new ManufacturerId($manufacturerId);
            $this->manufacturerId = $manufacturerId;
        } else {
            $this->manufacturerIdShouldBeReset = true;
        }
    }
```

Then in the handler, if `$command->shouldManufacturerIdBeReset` is true you reset it, else you ignore it.

And in order to increase DX, we can even a dedicated function (so people have 2 ways to reset the manufacturer ID: use regular setter with `null` or use this dedicated function)
```
public function setManufacturerIdToBeReset()
    {
        $this->manufacturerIdShouldBeReset = true;
    }
```